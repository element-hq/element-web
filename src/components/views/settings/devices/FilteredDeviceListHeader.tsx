/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { HTMLProps } from "react";
import { Tooltip } from "@vector-im/compound-web";

import { _t } from "../../../../languageHandler";
import StyledCheckbox, { CheckboxStyle } from "../../elements/StyledCheckbox";

interface Props extends Omit<HTMLProps<HTMLDivElement>, "className"> {
    selectedDeviceCount: number;
    isAllSelected: boolean;
    isSelectDisabled?: boolean;
    toggleSelectAll: () => void;
    children?: React.ReactNode;
}

const FilteredDeviceListHeader: React.FC<Props> = ({
    selectedDeviceCount,
    isAllSelected,
    isSelectDisabled,
    toggleSelectAll,
    children,
    ...rest
}) => {
    const checkboxLabel = isAllSelected ? _t("common|deselect_all") : _t("common|select_all");
    return (
        <div className="mx_FilteredDeviceListHeader" {...rest}>
            {!isSelectDisabled && (
                <Tooltip label={checkboxLabel} placement="top" isTriggerInteractive={false}>
                    <StyledCheckbox
                        kind={CheckboxStyle.Solid}
                        checked={isAllSelected}
                        onChange={toggleSelectAll}
                        id="device-select-all-checkbox"
                        data-testid="device-select-all-checkbox"
                        aria-label={checkboxLabel}
                    />
                </Tooltip>
            )}
            <span className="mx_FilteredDeviceListHeader_label">
                {selectedDeviceCount > 0
                    ? _t("settings|sessions|n_sessions_selected", { count: selectedDeviceCount })
                    : _t("settings|sessions|title")}
            </span>
            {children}
        </div>
    );
};

export default FilteredDeviceListHeader;
