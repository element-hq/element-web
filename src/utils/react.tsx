/*
Copyright 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import { ReactNode } from "react";
import { createRoot, Root } from "react-dom/client";

/**
 * Utility class to render & unmount additional React roots,
 * e.g. for pills, tooltips and other components rendered atop user-generated events.
 */
export class ReactRootManager {
    private roots: Root[] = [];
    private rootElements: Element[] = [];

    public get elements(): Element[] {
        return this.rootElements;
    }

    public render(children: ReactNode, element: Element): void {
        const root = createRoot(element);
        this.roots.push(root);
        this.rootElements.push(element);
        root.render(children);
    }

    public unmount(): void {
        while (this.roots.length) {
            const root = this.roots.pop()!;
            this.rootElements.pop();
            root.unmount();
        }
    }
}
