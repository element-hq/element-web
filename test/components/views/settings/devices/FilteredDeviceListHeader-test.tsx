/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { fireEvent, render } from "@testing-library/react";
import React from "react";

import FilteredDeviceListHeader from "../../../../../src/components/views/settings/devices/FilteredDeviceListHeader";

describe("<FilteredDeviceListHeader />", () => {
    const defaultProps = {
        selectedDeviceCount: 0,
        isAllSelected: false,
        toggleSelectAll: jest.fn(),
        children: <div>test</div>,
        ["data-testid"]: "test123",
    };
    const getComponent = (props = {}) => <FilteredDeviceListHeader {...defaultProps} {...props} />;

    it("renders correctly when no devices are selected", () => {
        const { container } = render(getComponent());
        expect(container).toMatchSnapshot();
    });

    it("renders correctly when all devices are selected", () => {
        const { container } = render(getComponent({ isAllSelected: true }));
        expect(container).toMatchSnapshot();
    });

    it("renders correctly when some devices are selected", () => {
        const { getByText } = render(getComponent({ selectedDeviceCount: 2 }));
        expect(getByText("2 sessions selected")).toBeTruthy();
    });

    it("clicking checkbox toggles selection", () => {
        const toggleSelectAll = jest.fn();
        const { getByTestId } = render(getComponent({ toggleSelectAll }));
        fireEvent.click(getByTestId("device-select-all-checkbox"));

        expect(toggleSelectAll).toHaveBeenCalled();
    });
});
