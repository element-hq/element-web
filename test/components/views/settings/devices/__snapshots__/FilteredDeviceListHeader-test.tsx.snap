// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilteredDeviceListHeader /> renders correctly when all devices are selected 1`] = `
<div>
  <div
    class="mx_FilteredDeviceListHeader"
    data-testid="test123"
  >
    <span
      tabindex="0"
    >
      <span
        class="mx_Checkbox mx_Checkbox_hasKind mx_Checkbox_kind_solid"
      >
        <input
          aria-label="Deselect all"
          checked=""
          data-testid="device-select-all-checkbox"
          id="device-select-all-checkbox"
          type="checkbox"
        />
        <label
          for="device-select-all-checkbox"
        >
          <div
            class="mx_Checkbox_background"
          >
            <div
              class="mx_Checkbox_checkmark"
            />
          </div>
        </label>
      </span>
    </span>
    <span
      class="mx_FilteredDeviceListHeader_label"
    >
      Sessions
    </span>
    <div>
      test
    </div>
  </div>
</div>
`;

exports[`<FilteredDeviceListHeader /> renders correctly when no devices are selected 1`] = `
<div>
  <div
    class="mx_FilteredDeviceListHeader"
    data-testid="test123"
  >
    <span
      tabindex="0"
    >
      <span
        class="mx_Checkbox mx_Checkbox_hasKind mx_Checkbox_kind_solid"
      >
        <input
          aria-label="Select all"
          data-testid="device-select-all-checkbox"
          id="device-select-all-checkbox"
          type="checkbox"
        />
        <label
          for="device-select-all-checkbox"
        >
          <div
            class="mx_Checkbox_background"
          >
            <div
              class="mx_Checkbox_checkmark"
            />
          </div>
        </label>
      </span>
    </span>
    <span
      class="mx_FilteredDeviceListHeader_label"
    >
      Sessions
    </span>
    <div>
      test
    </div>
  </div>
</div>
`;
